Gems/Terrain/Code/Source/TerrainRenderer/Passes/TerrainMacroTextureClipmapComputePass.cpp

```diff
@@ -0,0 +1,221 @@
+        }
+
+        {
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> colorPyramid = m_ownedAttachments[ClipmapStackSize];
+            AZ::RPI::Ptr<AZ::RPI::PassAttachment> normalPyramid = m_ownedAttachments[ClipmapStackSize * 2 + 1];
+            ProcessPassAttachment(pool, colorPyramid, m_colorClipmapPyramid);
+            ProcessPassAttachment(pool, normalPyramid, m_normalClipmapPyramid);
+        }
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::InitializeInternal()
+    {
+        for (uint32_t clipmapIndex = 0; clipmapIndex <= ClipmapStackSize; ++clipmapIndex)
+        {
+            m_clipmapData.SetPreviousClipmapCenter(AZ::Vector2(0.5f, 0.5f), clipmapIndex);
+            m_clipmapData.SetCurrentClipmapCenter(AZ::Vector2(0.5f, 0.5f), clipmapIndex);
+
+            m_clipmapData.m_fullUpdateFlag |= (1 << clipmapIndex);
+        }
+
+        m_clipmapData.m_previousViewPosition = AZ::Vector3::CreateZero();
+        m_clipmapData.m_currentViewPosition = AZ::Vector3::CreateZero();
+
+        m_colorClipmapStackIndex.Reset();
+        m_colorClipmapPyramidIndex.Reset();
+
+        ComputePass::InitializeInternal();
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::FrameBeginInternal(FramePrepareParams params)
+    {
+        UpdateClipmapData();
+
+        [[maybe_unused]] size_t x = sizeof(ClipmapData);
+
+        AZ::RPI::Scene* scene = m_pipeline->GetScene();
+        TerrainFeatureProcessor* terrainFeatureProcessor = scene->GetFeatureProcessor<TerrainFeatureProcessor>();
+        if (terrainFeatureProcessor)
+        {
+            const AZ::Aabb worldBounds = terrainFeatureProcessor->GetTerrainBounds();
+            m_clipmapData.m_worldBoundsMin = worldBounds.GetMin();
+            m_clipmapData.m_worldBoundsMax = worldBounds.GetMax();
+            m_clipmapData.SetMaxRenderSize(worldBounds.GetMax() - worldBounds.GetMin());
+        }
+        else
+        {
+            const AZ::Vector3 Zero = AZ::Vector3::CreateZero();
+            m_clipmapData.m_worldBoundsMin = Zero;
+            m_clipmapData.m_worldBoundsMax = Zero;
+            m_clipmapData.SetMaxRenderSize(Zero);
+        }
+
+        m_shaderResourceGroup->SetConstant(m_clipmapDataIndex, m_clipmapData);
+        m_clipmapData.m_fullUpdateFlag = 0;
+
+        ComputePass::FrameBeginInternal(params);
+    }
+
+    void TerrainMacroTextureClipmapGenerationPass::CompileResources(const AZ::RHI::FrameGraphCompileContext& context)
+    {
+        AZ::RPI::Scene* scene = m_pipeline->GetScene();
+        TerrainFeatureProcessor* terrainFeatureProcessor = scene->GetFeatureProcessor<TerrainFeatureProcessor>();
+        if (terrainFeatureProcessor)
+        {
+            auto terrainSrg = terrainFeatureProcessor->GetTerrainShaderResourceGroup();
+            if (terrainSrg)
+            {
+                if (!terrainSrg->IsQueuedForCompile())
+                {
+                    terrainSrg->Compile();
+                }
```

Comment:
I'm a little confused why the terrainSrg needs to be compiled every frame here. This class doesn't even make changes to the terrain srg, so it doesn't seem like it should be responsible for compiling it.

Reply:
Removed.